File: algorithms/algorithms-cplusplus/leetcode/Question_0682_Baseball_Game.cc
#include <string>
#include <vector>
#include <stack>

using namespace std;

class Solution {
 public:
  int calPoints(vector<string>& ops) {
    stack<int> s;
    for (string op : ops) {
      if (op == "+") {
        int a = s.top();
        s.pop();
        int b = a + s.top();
        s.push(a);
        s.push(b);
      } else if (op == "C") {
        s.pop();
      } else if (op == "D") {
        s.push(2 * s.top());
      } else {
        s.push(atoi(op.c_str()));
      }
    }
    int res = 0;
    while (!s.empty()) {
      res += s.top();
      s.pop();
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0682_Baseball_Game.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Stack;

public class Question_0682_Baseball_Game {
    public int calPoints(String[] ops) {
        Stack<Integer> stack = new Stack();
        for(String op : ops) {
            if (op.equals("+")) {
                int top = stack.pop();
                int newtop = top + stack.peek();
                stack.push(top);
                stack.push(newtop);
            } else if (op.equals("C")) {
                stack.pop();
            } else if (op.equals("D")) {
                stack.push(2 * stack.peek());
            } else {
                stack.push(Integer.valueOf(op));
            }
        }
        int res = 0;
        for(int score : stack) {
            res += score;
        }
        return res;
    }
}
